core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java

```diff
@@ -1042,6 +1042,10 @@ void testDyadicCollateOperator() {
         .columnType("VARCHAR(1) NOT NULL");
     expr("substring('a', 1, '3')")
         .columnType("VARCHAR(1) NOT NULL");
+
+    expr("SUBSTRING(NULL FROM 1 FOR 2)").ok();
```

Comment:
this test case is not effected by this issue, would you like to add a comment to clarify it?

Reply:
I just expand the tests coverage.

Reply:
I know it, if you add comment to clarify it, it would be better I think.

Reply:
@herunkang2018 fixed, can you make a review plz ? I update description and starting code point in appropriate issue.